### DISCUSSION TITLE
How to render a form to a chunk of HTML including integrations & JS without parent Twig context?

### DISCUSSION THREAD

**Discussion by rob-c-baker** [solspace/craft-freeform#1491] [Questions & Answers] [answered]
In an effort to make our site more cacheable I am trying to render forms in a PHP context with no parent Twig templates or layouts - the resulting HTML is delivered back to the client via an AJAX response.

This very nearly works. i.e. I get a form rendered in the HTML blob, but no scripts for things like the Freeform JS API or scripts to make ReCAPTCHAs work.

Code at the moment (simplified) is:

```php
$properties = [
  'fields' => [ ... ],
  'buttons' => [ ... ],
  'dynamicNotification' => [ ... ]
];
$form = Freeform::getInstance()->forms->getFormByHandle('form-handle');
$html = $form->render($properties);
```

I need to do something else to collect up all the JS scripts and other assets configured so I can deliver them back to the client. I have looked through the code and it looks like this (at least in part) happens via events.

So the question is: how do I get enough constituent elements out of Freeform, aside from the base HTML, from a "pure" PHP context to make a form work with the Freeform API & integrations like ReCAPTCHA?

Freeform version is `5.5.7` Craft version is `5.3.6` - both are Pro.

**_Edit:_** 

I should also add that I am injecting:

```html
<script type="text/javascript" src="https://www.google.com/recaptcha/api.js?render=explicit"></script>
```

Into the HTML as well, which is loading OK, but not being triggered I guess.

**Comment by rob-c-baker** [comment, marked as answer]
I have found a slightly janky way of making it work, the relevant part of the code in the controller now looks like this:

```php
$properties = [
  'fields' => [ ... ],
  'buttons' => [ ... ],
  'dynamicNotification' => [ ... ]
];
$form = Freeform::getInstance()->forms->getFormByHandle('form-handle');
$html = $form->render($properties);
$html .= (new FreeformVariable())->loadFormSpecificScripts($form);
$html .= '<script type="text/javascript" src="https://www.google.com/recaptcha/api.js?render=explicit"></script>';
```

Note the use of `(new FreeformVariable())->loadFormSpecificScripts($form);` which is supposed to be called from a Twig context that seems to do everything needed to produce the integration and API scripts needed.

**Comment by kjmartens** [comment]
Hi @rob-baker-ar,

Have you seen this latest addition in 5.5.4? 🙂 
https://docs.solspace.com/craft/freeform/v5/developer/js-plugin/#loading-freeform-js-manually

``` twig
{{ freeform.loadScripts([
    "freeform",
    "freeform.js-test",
    "recaptcha.v3",
    "stripe"
]) }}
```

**Reply by rob-c-baker** [reply]
This looks like a great addition.

In my case, it's nice to get just the scripts specific to the `form` - as configured in the CMS. As the code I have already accepts a form handle, `FreeformVariable::loadFormSpecificScripts()` is probably the best for now. The fact that the method only exists on the `FreeformVariable` class is not perfect but creates such a tiny overhead and it is something directly concerned with templating / HTML, it's not a problem.

Thanks for getting back to me.
